partiql-planner/src/main/kotlin/org/partiql/planner/internal/transforms/RexConverter.kt

```diff
@@ -148,73 +132,57 @@ internal object RexConverter {
-                    }
-                }
-                else -> pValue.type.toPType()
-            }
+            val datum = node.lit.toDatum()
+            val type = datum.type
             val cType = CompilerType(
                 _delegate = type,
                 isNullValue = node.lit.code() == Literal.NULL,
                 isMissingValue = node.lit.code() == Literal.MISSING
             )
-            val op = rexOpLit(pValue)
+            val op = rexOpLit(datum)
             return rex(cType, op)
         }
 
-        private fun Literal.toPartiQLValue(): PartiQLValue {
+        private fun Literal.toDatum(): Datum {
             val lit = this
             return when (lit.code()) {
-                Literal.NULL -> nullValue()
-                Literal.MISSING -> missingValue()
-                Literal.STRING -> stringValue(lit.stringValue())
-                Literal.BOOL -> boolValue(lit.booleanValue())
+                Literal.NULL -> Datum.nullValue()
+                Literal.MISSING -> Datum.missing()
+                Literal.STRING -> Datum.string(lit.stringValue())
+                Literal.BOOL -> Datum.bool(lit.booleanValue())
                 Literal.EXACT_NUM -> {
-                    // TODO previous behavior inferred decimals with scale = 0 to be a PartiQLValue.IntValue with
-                    //  PType of numeric. Since we're keeping numeric and decimal, need to take another look at
-                    //  whether the literal should have type decimal or numeric.
                     val dec = lit.bigDecimalValue().round(MathContext(38, RoundingMode.HALF_EVEN))
-                    if (dec.scale() == 0) {
-                        intValue(dec.toBigInteger())
-                    } else {
-                        decimalValue(dec)
-                    }
+                    Datum.decimal(dec, dec.precision(), dec.scale())
                 }
                 Literal.INT_NUM -> {
                     val n = lit.numberValue()
                     // 1st, try parse as int
                     try {
                         val v = n.toInt(10)
-                        return int32Value(v)
+                        return Datum.integer(v)
                     } catch (ex: NumberFormatException) {
                         // ignore
                     }
 
                     // 2nd, try parse as long
                     try {
                         val v = n.toLong(10)
-                        return int64Value(v)
+                        return Datum.bigint(v)
                     } catch (ex: NumberFormatException) {
                         // ignore
                     }
 
                     // 3rd, try parse as BigInteger
                     try {
                         val v = BigInteger(n)
-                        return intValue(v)
+                        val vDecimal = BigDecimal(v)
+                        return Datum.decimal(vDecimal, vDecimal.precision(), vDecimal.scale())
```

Comment:
Farewell arbitrary integers. This makes me think creating a decimal datum from a BigInteger is a reasonable API.